### PR TITLE
Benchmarks

### PR DESCRIPTION
**Pull request by N-Maas**
This branch is intended to collect some ideas for benchmarks.

If you want to see a really huge difference between `-O1` and `-O0`, check out the `AbstractedMath` benchmark :smile:

### CODE REVIEW

File: mjtest-files/compile/bench/SubsetSum.mj [hunk 1]
```diff
@@ -0,0 +1,193 @@
+class A {
```
Comment (larsk21) [inline mjtest-files/compile/bench/SubsetSum.mj:1]: Could you add a short description what is tested here, similar to the other benchmarks?

**Comment by larsk21** [comment]
> changed the base branch from f/final to master

Is there a reason for this?

**Comment by N-Maas** [comment]
> > changed the base branch from f/final to master
> 
> Is there a reason for this?

Mostly that I haven't updated `f/final` yet

File: mjtest-files/compile/bench/SubsetSum.mj [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+    public int runSingle(int target, int n) {
+        int row = 0;
+        while (row < num_rows) {
+            int i = 0;
+            while (!permutations[row][i]) {
+                i = i + 1;
+            }
+            int a = values[i];
+            i = i + 1;
```
Comment (0xNB) [inline mjtest-files/compile/bench/SubsetSum.mj:34]: couldnt you declare new variables here to increase register pressure?

File: mjtest-files/compile/bench/SubsetSum.mj [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        while (n_a < 2) {
+            int n_b = 0;
+            while (n_b < 2) {
+                int n_c = 0;
+                while (n_c < 2) {
+                    int n_d = 0;
+                    while (n_d < 2) {
+                        int sum = n_a * a + n_b * b + n_c * c + n_d * d;
+                        if (sum == target) {
+                            n_solutions = n_solutions + 1;
```
Comment (0xNB) [inline mjtest-files/compile/bench/SubsetSum.mj:180]: this is bad code, we really need some refactoring here that adds a new method `addOne(n_solutions)` to make this part more readable

File: mjtest-files/compile/bench/AbstractedMath.mj [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    public int log(int base, int val) {
+        int sign = 1;
+        if (val < 0) {
+            sign = -1;
+            val = -val;
+        }
+        int sum = 0;
+        int guess = 5;
+        while (val / pow(base, guess) > 1) {
+            sum = sum + guess;
```
Comment (0xNB) [inline mjtest-files/compile/bench/AbstractedMath.mj:38]: needs some refactoring to a method that performs an addition, to make code more readable

File: mjtest-files/compile/bench/SubsetSum.mj (hunk 1 above)
Comment (larsk21) [reply mjtest-files/compile/bench/SubsetSum.mj]: Thanks!

File: mjtest-files/compile/bench/SubsetSum.mj (hunk 2 above)
Comment (N-Maas) [reply mjtest-files/compile/bench/SubsetSum.mj:37]: That's not the purpose of the benchmark, and declaring new variables doesn't help anyway if they are just optimized away in the middle end.

File: mjtest-files/compile/bench/SubsetSum.mj [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            permutations[i] = new boolean[n];
+            i = i + 1;
+        }
+        initSubsets(new boolean[0], 0, 6, 0, n);
+
+        solutionsPerValue = new int[n];
+    }
+
+    /* prefix.length == pos */
+    public int initSubsets(boolean[] prefix, int pos, int remaining, int row, int n) {
```
Comment (0xNB) [inline mjtest-files/compile/bench/SubsetSum.mj:136]: this method really needs some unused parameters ...

File: mjtest-files/compile/bench/AbstractedMath.mj (hunk 4 above)
Comment (N-Maas) [reply mjtest-files/compile/bench/AbstractedMath.mj:38]: Don't know. A compiler without an inliner doesn't have the slightest chance here, anyway

**Pull request merged**